### PR TITLE
Added props to disable BlurGate

### PR DESCRIPTION
**Pull request by PeterKottas**
Detailed in https://github.com/react-page/react-page/issues/800

### CODE REVIEW

**Comment by PeterKottas** [comment]
Is the CI expected to fail? Cause this doesn't seem like something I touched.

**Comment by macrozone** [comment]
> Is the CI expected to fail? Cause this doesn't seem like something I touched.

do you see the ci? the linting fails as you changed from single quotes to double quotes. Please check your editor, the project should contain some linting/prettier rules (or i hope so?)

**Comment by PeterKottas** [comment]
Weird, it always worked out of the box. Maybe I have different defaults on this machine.

**Comment by PeterKottas** [comment]
Mate, looking at https://github.com/react-page/react-page/blob/master/tslint.json#L67, I think it should be single. My machine agrees. So not sure why the CI is coughing up. Also, it's single in most other files.

**Comment by macrozone** [comment]
> Mate, looking at https://github.com/react-page/react-page/blob/master/tslint.json#L67, I think it should be single. My machine agrees. So not sure why the CI is coughing up. Also, it's single in most other files.
> 
> > the linting fails as you changed from single quotes to double quotes
> 
> I think fact changed from double to single which should be correct according to tslint config.

yes it should be single, but you changed it to double quotes

File: package.json [hunk 1]
```diff
@@ -61,7 +61,7 @@
     "lint:watch": "npm-run-all watch:ts -- -c \"npm-run-all lint\"",
     "prettier": " prettier-tslint fix \"packages/**/src/**/*.(ts|tsx)\"",
     "publish": "lerna publish",
-    "start": "TSC_WATCHFILE=UseFsEventsWithFallbackDynamicPolling npm-run-all -p build:watch start:examples",
+    "start": "cross-env TSC_WATCHFILE=UseFsEventsWithFallbackDynamicPolling npm-run-all -p build:watch start:examples",
```
Comment (PeterKottas) [inline package.json:64]: This is unrelated, just something that must have been added a while ago and it wasn't done properly (cross platform)

File: packages/core/src/actions/cell/insert.ts [hunk 2]
```diff
@@ -129,6 +129,14 @@ export const insertCellRightInline = insert(CELL_INSERT_INLINE_RIGHT);
...
   return {
     ...item,
+    content: item.content && {
+      plugin: item.content.plugin,
+      state: JSON.parse(JSON.stringify(item.content.state)),
+    },
+    layout: item.layout && {
+      plugin: item.layout.plugin,
+      state: JSON.parse(JSON.stringify(item.layout.state)),
+    },
```
Comment (PeterKottas) [inline packages/core/src/actions/cell/insert.ts:139]: This is what fixes the duplicate button problem mentioned in #802 . Sorry again about mixing PRs.

File: packages/core/src/Provider/index.tsx [hunk 3]
```diff
@@ -5,9 +5,12 @@ import HTML5Backend from 'react-dnd-html5-backend';
 import BlurGate from '../components/BlurGate';
 import Editor from '../Editor';
 import { ReduxProvider } from '../reduxConnect';
+import { DisplayModes } from '../actions/display';
 interface ProviderProps {
   editor: Editor;
   dndBackend?: BackendFactory;
+  blurGateDisabled?: boolean;
+  blurGateDefaultMode?: DisplayModes;
```
Comment (macrozone) [inline packages/core/src/Provider/index.tsx:13]: we should probably also add this in `Editor` from 'react-page/editor` as its now recommended to use just this simple component. 

but there i would name it `displayModeDefault` or `defaultDisplayMode`

do you still need the `blurGateDisabled` ? is there another issue with it?

File: package.json (hunk 1 above)
Comment (macrozone) [reply package.json:64]: perfect

File: packages/core/src/actions/cell/insert.ts (hunk 2 above)
Comment (macrozone) [reply packages/core/src/actions/cell/insert.ts:139]: no worries :-) looks good. This would probably fail if some state happens to be circular (which should not be the case in my opinion)

File: packages/core/src/Provider/index.tsx (hunk 3 above)
Comment (PeterKottas) [reply packages/core/src/Provider/index.tsx:13]: Yp I'll add it there. Yeah, I need to disable it. Even the blurAllCells doesn't work for me. I don't want my bottom toolbar closing while I am clicking inside a different portal. ```defaultDisplayMode``` sounds good. I will also add ```blurGateDisabled``` there. K?

**Comment by PeterKottas** [comment]
> yes it should be single, but you changed it to double quotes

My bad, sry got myself confused. I was making the code changes without having the node_modules installed. That's why it used some vs-code defaults instead of correct tslint settings. All good now.

File: packages/core/src/actions/cell/insert.ts (hunk 2 above)
Comment (PeterKottas) [reply packages/core/src/actions/cell/insert.ts:139]: Agreed! Cheers

**Comment by PeterKottas** [comment]
ok, everything we discussed so far has been implemented.

**Comment by PeterKottas** [comment]
Seeing that the 2 of us are for the most part the most active contributors/maintainers, I was curious if it would make sense to discuss our approach to closing and merging PRs. Naturally, it's the best solution for us to check each other's work as it should be, but I am sure I've previously merged in my own PRs and so have you. It simply cannot be any other way given that we are not doing this full time and therefore have periods of time where we cannot attend to this library.

I was thinking we should probably create a few simple rules on how to proceed to make our lives easiest and most productive.

1. Let's give each other 24 hours to react to a new PR.
2. If the other person doesn't react within 24 hours we should be free to close our own PRs without further "second pair of eyes" with the exception of:
a) the PR is huge (naturally this is on case to case bases, I think we are capable enough to find the right balance here)
b) the PR proposes a breaking change to the API 
c) the PR touches something that can be considered a matter of preference

If any of these are the case, we should allow further 48 hours (even if there is no response for the other person to pass). Understandably, it's preferable to communicate with each other and for instance say "I cannot look at it for the next 3 days, but please wait for me, I'll find the time during the weekend" or something along those lines. These "guidelines" are mostly meant for when one of us is unreachable for one reason or another. Also simple stuff like style/lint issues, docs or comments fixes, cross-platform issues etc. should be merged at the person's discretion without having to wait at all.

What do you think?

**Comment by PeterKottas** [comment]
Could you pls take a look @macrozone ? I've got some other stuff riding on this, cheers!

File: packages/core/src/Provider/index.tsx (hunk 3 above)
Comment (macrozone) [reply packages/core/src/Provider/index.tsx:13]: @PeterKottas the blur should work through a portal. in fact, the bottom toolbar is a portal.

The BlurGate checks whether the click was inside the react-tree of the component.
But this has only been tested with the @react-page/editor, not with the old approach. I think i now why it is not working

**Comment by macrozone** [comment]
> Seeing that the 2 of us are for the most part the most active contributors/maintainers, I was curious if it would make sense to discuss our approach to closing and merging PRs. Naturally, it's the best solution for us to check each other's work as it should be, but I am sure I've previously merged in my own PRs and so have you. It simply cannot be any other way given that we are not doing this full time and therefore have periods of time where we cannot attend to this library.
> 
> I was thinking we should probably create a few simple rules on how to proceed to make our lives easiest and most productive.
> 
> 1. Let's give each other 24 hours to react to a new PR.
> 2. If the other person doesn't react within 24 hours we should be free to close our own PRs without further "second pair of eyes" with the exception of:
>    a) the PR is huge (naturally this is on case to case bases, I think we are capable enough to find the right balance here)
>    b) the PR proposes a breaking change to the API
>    c) the PR touches something that can be considered a matter of preference
> 
> If any of these are the case, we should allow further 48 hours (even if there is no response for the other person to pass). Understandably, it's preferable to communicate with each other and for instance say "I cannot look at it for the next 3 days, but please wait for me, I'll find the time during the weekend" or something along those lines. These "guidelines" are mostly meant for when one of us is unreachable for one reason or another. Also simple stuff like style/lint issues, docs or comments fixes, cross-platform issues etc. should be merged at the person's discretion without having to wait at all.
> 
> What do you think?

I  totally agree to move fast forward. and agree with the reaction time ;-)

In this case i am fine if the `disableBlurGate` prop stays, but there are two things to consider:

- I still would like to know why it does not work properly for you, but we can investigate after we have merged it
- we should reduce the api surface, so every new prop should be challenged. Maybe we define that we "underscore" non-documented, possible future breaking props like this? `_disableBlurGate`, so that we don't have to apologize in the future ;-)

also notice that if you disable the blur gates functionality, your users can get confused with a wrong plugin-panel

**Comment by macrozone** [comment]
you can merge if you want, notice my concerns above

**Comment by PeterKottas** [comment]
Cool, I will proceed accordingly. 

> I still would like to know why it does not work properly for you, but we can investigate after we have merged it

Simply put, I need the plugin panel to stay open even after I click away. In my tutorial that uses https://reactour.js.org/, I am highlighting parts of the bottomToolbar and letting people click next to move to the next button explanation. The problem is, if I click next, the toolbar disappears breaking the functionality. 

> Agreed about the API surface and the props, although I must say, right now, I think we lack behind in terms of customization. The only way to achieve that is to somewhat increase the API surface. Sensibly of course.

**Comment by PeterKottas** [comment]
Quick question, is all that's needed to release new version pushing a new tag? Rest is automated? I am not 100% sure. TY
